common/src/unifyfs_client_rpcs.h

```diff
@@ -282,6 +283,109 @@ MERCURY_GEN_PROC(unifyfs_heartbeat_in_t,
 MERCURY_GEN_PROC(unifyfs_heartbeat_out_t, ((int32_t)(ret)))
 DECLARE_MARGO_RPC_HANDLER(unifyfs_heartbeat_rpc)
 
+/* unifyfs_node_local_extents_get_rpc (client => server)
+ *
+ * returns node local extents for a file
+ * given a global file id */
+MERCURY_GEN_STRUCT_PROC(unifyfs_client_index_t,
+                 ((uint64_t)(file_pos))
+                 ((uint64_t)(log_pos))
+                 ((uint64_t)(length))
+                 ((int32_t)(gfid))
+                 ((int32_t)(log_app_id))
+                 ((int32_t)(log_client_id)))
+/* List of extents used in the output of RPC call*/
+typedef struct extents_list {
+    unifyfs_client_index_t value;
+    struct extents_list* next;
+} extents_list_t;
```

Comment:
I guess I'm still a little confused by your choice of a list-of-structures here, rather than a simple array-of-structures. Seems like it adds a bunch of extra encoding/decoding overhead versus just transferring the bulk array and using simple indexing.

Reply:
So here, I need to send a list of extents on input as well as the output. It was more efficient to transfer it directly as its only metadata.

Reply:
I think you misunderstood my suggestion. If you just use an array of `unifyfs_client_index_t`, you don't need to write the custom serialization routine and do all the list pointer work. See for example `unifyfs_mread_rpc` where we pass an array of `unifyfs_extent_t` as bulk data.

Reply:
I can use the bulk data while sending unifyfs_client_index_t but the return doesn't work for me. I am forced to write another RPC from server to client like in case of mread_rpc.

Can you point me to some documentation on how I can return bulk data from an RPC?

Reply:
For client-server RPCs, it's all going through shared memory, not using RDMA. My experience is that put/get of whole arrays in shmem is way faster than iterating over a list where you encode/decode each list item.

The size of the array can be dynamic, you only need it to calculate the bulk size (`array size = # elements * sizeof(element)`).

To return bulk data, you just add the bulk args (size and data) to the output struct member list in the `MERCURY_GEN_PROC()`, same as is done for the input struct.

Reply:
Ok, I actually missed pulling data from the request as well as response bulk data that's why it was not working for me.

I fixed that. I had to move the pull function to the common rpc_util file.

I have done the changes and tested it as well